lib/Analysis/Membar.cpp

```diff
@@ -64,29 +60,74 @@ void MembarAnalysis::resolve(FunctionOpInterface funcOp,
     // so overwrite the output state entirely.
     outputBlockInfoMap[block] = inputBlockInfo;
     // Update the successors
-    for (auto *successor : successors) {
+    for (VirtualBlock successor : successors) {
       inputBlockInfoMap[successor].join(outputBlockInfoMap[block]);
       blockList.emplace_back(successor);
     }
   }
 
   // Update the final dangling buffers that haven't been synced
-  auto &funcBlockInfo = (*funcBlockInfoMap)[funcOp];
-  funcOp.walk<WalkOrder::PreOrder>([&](Block *block) {
-    block->walk([&](triton::ReturnOp returnOp) {
-      funcBlockInfo.join(outputBlockInfoMap[block]);
+  BlockInfo &funcBlockInfo = (*funcBlockInfoMap)[funcOp];
+  funcOp.walk<WalkOrder::PreOrder>([&](triton::ReturnOp returnOp) {
+    // Find the virtual block that contains this op.
+    SmallVector<std::pair<VirtualBlock, BlockInfo>> virtualBlocks;
+    for (auto &[block, blockInfo] : outputBlockInfoMap) {
+      if (block.first == returnOp->getBlock())
+        virtualBlocks.emplace_back(block, blockInfo);
+    }
+    // The last virtual block in this group contains the terminator.
```

Comment:
Can you define what `last` mean here? The block that belongs to the outmost region?

Reply:
The one that actually contains the terminator. I'll clarify this.